// material-ui/TextField/TextFieldLabel.js
'use strict';
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _simpleAssign = require('simple-assign');
var _simpleAssign2 = _interopRequireDefault(_simpleAssign);
var _react = require('react');
var _react2 = _interopRequireDefault(_react);
var _transitions = require('../styles/transitions');
var _transitions2 = _interopRequireDefault(_transitions);
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function getStyles(props) {
    var defaultStyles = {
        position: 'absolute',
        lineHeight: '22px',
        top: 38,
        transition: _transitions2.default.easeOut(),
        zIndex: 1,
        transform: 'scale(1) translate(0, 0)',
        transformOrigin: 'left top',
        pointerEvents: 'auto',
        userSelect: 'none'
    };
    var shrinkStyles = props.shrink ? (0, _simpleAssign2.default)({
        transform: 'scale(0.75) translate(0, -28px)',
        pointerEvents: 'none'
    }, props.shrinkStyle) : null;
    return {
        root: (0, _simpleAssign2.default)(defaultStyles, props.style, shrinkStyles)
    };
}
var TextFieldLabel = function TextFieldLabel(props) {
    var muiTheme = props.muiTheme, className = props.className, children = props.children, htmlFor = props.htmlFor, onTouchTap = props.onTouchTap;
    var prepareStyles = muiTheme.prepareStyles;
    var styles = getStyles(props);
    return _react2.default.createElement('label', {
        className: className,
        style: prepareStyles(styles.root),
        htmlFor: htmlFor,
        onTouchTap: onTouchTap
    }, children);
};
process.env.NODE_ENV !== "production" ? TextFieldLabel.propTypes = {
    children: _react.PropTypes.node,
    className: _react.PropTypes.string,
    disabled: _react.PropTypes.bool,
    htmlFor: _react.PropTypes.string,
    muiTheme: _react.PropTypes.object.isRequired,
    onTouchTap: _react.PropTypes.func,
    shrink: _react.PropTypes.bool,
    shrinkStyle: _react.PropTypes.object,
    style: _react.PropTypes.object
} : void 0;
TextFieldLabel.defaultProps = {
    disabled: false,
    shrink: false
};
exports.default = TextFieldLabel;
