// mobx-react-form/lib/Events.js
'use strict';
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _intersection2 = require('lodash/intersection');
var _intersection3 = _interopRequireDefault(_intersection2);
var _omitBy2 = require('lodash/omitBy');
var _omitBy3 = _interopRequireDefault(_omitBy2);
var _keys2 = require('lodash/keys');
var _keys3 = _interopRequireDefault(_keys2);
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor)
        descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
} } return function (Constructor, protoProps, staticProps) { if (protoProps)
    defineProperties(Constructor.prototype, protoProps); if (staticProps)
    defineProperties(Constructor, staticProps); return Constructor; }; }();
var _desc, _value, _class, _descriptor;
var _mobx = require('mobx');
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _initDefineProp(target, property, descriptor, context) {
    if (!descriptor)
        return;
    Object.defineProperty(target, property, {
        enumerable: descriptor.enumerable,
        configurable: descriptor.configurable,
        writable: descriptor.writable,
        value: descriptor.initializer ? descriptor.initializer.call(context) : void 0
    });
}
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
} }
function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
    var desc = {};
    Object['ke' + 'ys'](descriptor).forEach(function (key) {
        desc[key] = descriptor[key];
    });
    desc.enumerable = !!desc.enumerable;
    desc.configurable = !!desc.configurable;
    if ('value' in desc || desc.initializer) {
        desc.writable = true;
    }
    desc = decorators.slice().reverse().reduce(function (desc, decorator) {
        return decorator(target, property, desc) || desc;
    }, desc);
    if (context && desc.initializer !== void 0) {
        desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
        desc.initializer = undefined;
    }
    if (desc.initializer === void 0) {
        Object['define' + 'Property'](target, property, desc);
        desc = null;
    }
    return desc;
}
function _initializerWarningHelper(descriptor, context) {
    throw new Error('Decorating class property failed. Please ensure that transform-class-properties is enabled.');
}
var Events = (_class = function () {
    function Events() {
        _classCallCheck(this, Events);
        _initDefineProp(this, '$running', _descriptor, this);
        this.$path = {
            clear: null,
            reset: null,
            update: null,
            validate: null
        };
    }
    _createClass(Events, [{
            key: 'getRunning',
            value: function getRunning() {
                var key = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
                if (key)
                    return this.$running[key];
                return this.$running;
            }
        }, {
            key: 'setRunning',
            value: function setRunning(key, flag) {
                var path = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
                this.$running[key] = flag;
                if (path)
                    this.$path[key] = path;
                else
                    this.$path[key] = null;
            }
        }, {
            key: 'path',
            value: function path(key) {
                return this.$path[key];
            }
        }, {
            key: 'running',
            value: function running(events) {
                var running = (0, _keys3.default)((0, _omitBy3.default)(this.getRunning(), function (e) {
                    return e === false;
                }));
                return (0, _intersection3.default)(events, running).length > 0;
            }
        }]);
    return Events;
}(), (_descriptor = _applyDecoratedDescriptor(_class.prototype, '$running', [_mobx.observable], {
    enumerable: true,
    initializer: function initializer() {
        return {
            clear: false,
            reset: false,
            update: false,
            validate: false
        };
    }
}), _applyDecoratedDescriptor(_class.prototype, 'setRunning', [_mobx.action], Object.getOwnPropertyDescriptor(_class.prototype, 'setRunning'), _class.prototype)), _class);
exports.default = new Events();
module.exports = exports['default'];
