// mobx-react-form/lib/Options.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor)
        descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
} } return function (Constructor, protoProps, staticProps) { if (protoProps)
    defineProperties(Constructor.prototype, protoProps); if (staticProps)
    defineProperties(Constructor, staticProps); return Constructor; }; }();
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
} }
var Options = function () {
    function Options() {
        _classCallCheck(this, Options);
        this.options = {
            showErrorsOnInit: false,
            validateOnInit: true,
            validateOnChange: true,
            strictUpdate: false,
            strictDelete: true,
            showErrorsOnUpdate: true,
            alwaysShowDefaultError: false,
            defaultGenericError: null,
            loadingMessage: null,
            allowRequired: false,
            autoParseNumbers: false
        };
    }
    _createClass(Options, [{
            key: "get",
            value: function get() {
                var key = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
                if (key)
                    return this.options[key];
                return this.options;
            }
        }, {
            key: "set",
            value: function set(options) {
                Object.assign(this.options, options);
            }
        }]);
    return Options;
}();
exports.default = Options;
module.exports = exports["default"];
