// mobx-react-form/lib/utils.js
'use strict';
Object.defineProperty(exports, "__esModule", {
    value: true
});
var _uniqueId2 = require('lodash/uniqueId');
var _uniqueId3 = _interopRequireDefault(_uniqueId2);
var _isUndefined2 = require('lodash/isUndefined');
var _isUndefined3 = _interopRequireDefault(_isUndefined2);
var _max2 = require('lodash/max');
var _max3 = _interopRequireDefault(_max2);
var _isInteger2 = require('lodash/isInteger');
var _isInteger3 = _interopRequireDefault(_isInteger2);
var _ary2 = require('lodash/ary');
var _ary3 = _interopRequireDefault(_ary2);
var _map2 = require('lodash/map');
var _map3 = _interopRequireDefault(_map2);
var _has2 = require('lodash/has');
var _has3 = _interopRequireDefault(_has2);
var _partial2 = require('lodash/partial');
var _partial3 = _interopRequireDefault(_partial2);
var _trimEnd2 = require('lodash/trimEnd');
var _trimEnd3 = _interopRequireDefault(_trimEnd2);
var _replace2 = require('lodash/replace');
var _replace3 = _interopRequireDefault(_replace2);
var _isString2 = require('lodash/isString');
var _isString3 = _interopRequireDefault(_isString2);
var _isArray2 = require('lodash/isArray');
var _isArray3 = _interopRequireDefault(_isArray2);
var _isNil2 = require('lodash/isNil');
var _isNil3 = _interopRequireDefault(_isNil2);
var _intersection2 = require('lodash/intersection');
var _intersection3 = _interopRequireDefault(_intersection2);
var _union2 = require('lodash/union');
var _union3 = _interopRequireDefault(_union2);
var _every2 = require('lodash/every');
var _every3 = _interopRequireDefault(_every2);
var _some2 = require('lodash/some');
var _some3 = _interopRequireDefault(_some2);
var _typeof = typeof Symbol === "function" && typeof Symbol.iterator === "symbol" ? function (obj) { return typeof obj; } : function (obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; };
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var computed = ['hasError', 'isValid', 'isDirty', 'isPristine', 'isDefault', 'isEmpty', 'focus', 'touched', 'changed', 'disabled'];
var props = ['value', 'initial', 'default', 'label', 'placeholder', 'disabled', 'related', 'options', 'bindings', 'type', 'error'];
var iprops = ['values', 'initials', 'defaults', 'labels', 'placeholders', 'disabled', 'related', 'options', 'bindings', 'types'];
var vprops = ['rules', 'validate'];
var check = function check(_ref) {
    var type = _ref.type, data = _ref.data;
    var $check = void 0;
    switch (type) {
        case 'some':
            $check = function $check($data) {
                return (0, _some3.default)($data, Boolean);
            };
            break;
        case 'every':
            $check = function $check($data) {
                return (0, _every3.default)($data, Boolean);
            };
            break;
        default:
            $check = null;
    }
    return $check(data);
};
var has = function has($type, $data) {
    var $ = void 0;
    switch ($type) {
        case 'props':
            $ = props;
            break;
        case 'computed':
            $ = computed;
            break;
        case 'all':
            $ = (0, _union3.default)(computed, props, vprops);
            break;
        default:
            $ = null;
    }
    return (0, _intersection3.default)($data, $).length > 0;
};
var allowed = function allowed(type, data) {
    if (has(type, data))
        return;
    var $msg = 'The selected property is not allowed';
    throw new Error($msg + ' (' + JSON.stringify(data) + ')');
};
var throwError = function throwError(path, fields) {
    var msg = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
    if (!(0, _isNil3.default)(fields))
        return;
    var $msg = (0, _isNil3.default)(msg) ? 'The selected field is not defined' : msg;
    throw new Error($msg + ' (' + path + ')');
};
var isPromise = function isPromise(obj) {
    return !!obj && typeof obj.then === 'function' && ((typeof obj === 'undefined' ? 'undefined' : _typeof(obj)) === 'object' || typeof obj === 'function');
};
var isEvent = function isEvent(obj) {
    if (typeof Event === 'undefined')
        return false;
    return obj instanceof Event || !(0, _isNil3.default)(obj.target);
};
var isStruct = function isStruct(data) {
    return (0, _isArray3.default)(data) && (0, _every3.default)(data, _isString3.default);
};
var pathToStruct = function pathToStruct(path) {
    var struct = void 0;
    struct = (0, _replace3.default)(path, new RegExp('[.]\\d($|.)', 'g'), '[].');
    struct = (0, _replace3.default)(struct, '..', '.');
    struct = (0, _trimEnd3.default)(struct, '.');
    return struct;
};
var hasSome = function hasSome(obj, keys) {
    return (0, _some3.default)(keys, (0, _partial3.default)(_has3.default, obj));
};
var hasUnifiedProps = function hasUnifiedProps(field) {
    return hasSome(field, props) || hasSome(field, vprops);
};
var hasSeparatedProps = function hasSeparatedProps(initial) {
    return hasSome(initial, iprops) || hasSome(initial, vprops);
};
var parseIntKeys = function parseIntKeys(fields) {
    return (0, _map3.default)(fields.keys(), (0, _ary3.default)(parseInt, 1));
};
var hasIntKeys = function hasIntKeys(fields) {
    return (0, _every3.default)(parseIntKeys(fields), _isInteger3.default);
};
var maxKey = function maxKey(fields) {
    var max = (0, _max3.default)(parseIntKeys(fields));
    return (0, _isUndefined3.default)(max) ? 0 : max;
};
var makeId = function makeId(path) {
    return (0, _uniqueId3.default)([(0, _replace3.default)(path, new RegExp('\\.', 'g'), '-'), '--'].join(''));
};
exports.default = {
    computed: computed,
    props: props,
    iprops: iprops,
    vprops: vprops,
    check: check,
    has: has,
    allowed: allowed,
    throwError: throwError,
    isPromise: isPromise,
    isEvent: isEvent,
    isStruct: isStruct,
    pathToStruct: pathToStruct,
    hasUnifiedProps: hasUnifiedProps,
    hasSeparatedProps: hasSeparatedProps,
    parseIntKeys: parseIntKeys,
    hasIntKeys: hasIntKeys,
    maxKey: maxKey,
    makeId: makeId
};
module.exports = exports['default'];
